Internal MSBuild Error during logging by AssemblyLoadsTracker

### Issue Description


This is with 10.0.100-alpha.1.24573.1
Build log while it lasts https://[dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/887239/logs/226](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/887239/logs/226)

I thought this might be a task bug similar to the Restore task bug that caused https://github.com/dotnet/msbuild/issues/4792. Then wondered, how I could make it log which task was logging. Then realized the logging seems to be from within "AssemblyLoadsTracker" not a task.

```
Build FAILED.

MSBUILD : error MSB4166: Child node "4" exited prematurely. Shutting down. Diagnostic information may be found in files in "/mnt/vss/_work/1/s/artifacts/log/Release" and will be named MSBuild_*.failure.txt. This location can be changed by setting the MSBUILDDEBUGPATH environment variable to a different directory. [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: /mnt/vss/_work/1/s/artifacts/log/Release/MSBuild_pid-21939_4a3f1338ba4d49eeb5ba3df414c5b8de.failure.txt: [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: UNHANDLED EXCEPTIONS FROM PROCESS 21939: [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: ===================== [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: 12/5/2024 3:07:31 PM [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: Microsoft.Build.Framework.InternalErrorException: MSB0001: Internal MSBuild Error: ContextID 1702 should have been in the ID-to-project file mapping but wasn't! [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: at Microsoft.Build.Shared.ErrorUtilities.ThrowInternalError(String message, Object[] args) [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: at Microsoft.Build.BackEnd.Logging.LoggingService.GetAndVerifyProjectFileFromContext(BuildEventContext context) [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: at Microsoft.Build.BackEnd.Logging.LoggingService.LogBuildEvent(BuildEventArgs buildEvent) [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: at Microsoft.Build.BackEnd.Components.RequestBuilder.AssemblyLoadsTracker.CurrentDomainOnAssemblyLoad(Object sender, AssemblyLoadEventArgs args) [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: =================== [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166:  [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166:  [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: Child node "4" exited prematurely. Shutting down. Diagnostic information may be found in files in "/mnt/vss/_work/1/s/artifacts/log/Release" and will be named MSBuild_*.failure.txt. This location can be changed by setting the MSBUILDDEBUGPATH environment variable to a different directory. [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: /mnt/vss/_work/1/s/artifacts/log/Release/MSBuild_pid-21939_4a3f1338ba4d49eeb5ba3df414c5b8de.failure.txt: [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: UNHANDLED EXCEPTIONS FROM PROCESS 21939: [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: ===================== [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: 12/5/2024 3:07:31 PM [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: Microsoft.Build.Framework.InternalErrorException: MSB0001: Internal MSBuild Error: ContextID 1702 should have been in the ID-to-project file mapping but wasn't! [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: at Microsoft.Build.Shared.ErrorUtilities.ThrowInternalError(String message, Object[] args) [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: at Microsoft.Build.BackEnd.Logging.LoggingService.GetAndVerifyProjectFileFromContext(BuildEventContext context) [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: at Microsoft.Build.BackEnd.Logging.LoggingService.LogBuildEvent(BuildEventArgs buildEvent) [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: at Microsoft.Build.BackEnd.Components.RequestBuilder.AssemblyLoadsTracker.CurrentDomainOnAssemblyLoad(Object sender, AssemblyLoadEventArgs args) [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166: =================== [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166:  [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
MSBUILD : error MSB4166:  [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
EXEC : npm error Lifecycle script `integration-test` failed with error :  [/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj]
/mnt/vss/_work/1/s/eng/Npm.Workspace.FunctionalTests.nodeproj(28,5): error MSB3073: The command "npm run integration-test" exited with code 1.
```

### Steps to Reproduce

Don't know how to repro. Binary log on request, but I think this may start with code inspection

### Expected Behavior

.

### Actual Behavior

.

### Analysis

_No response_

### Versions & Configurations

_No response_

--- a/src/Build/BackEnd/Components/Logging/LoggingService.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingService.cs
@@ -1252,15 +1252,18 @@ namespace Microsoft.Build.BackEnd.Logging
 
             if ((warningEvent = buildEvent as BuildWarningEventArgs) != null && warningEvent.BuildEventContext != null && warningEvent.BuildEventContext.ProjectContextId != BuildEventContext.InvalidProjectContextId)
             {
-                warningEvent.ProjectFile = GetAndVerifyProjectFileFromContext(warningEvent.BuildEventContext);
+                warningEvent.ProjectFile = GetAndVerifyProjectFileFromContext(warningEvent, false);
             }
             else if ((errorEvent = buildEvent as BuildErrorEventArgs) != null && errorEvent.BuildEventContext != null && errorEvent.BuildEventContext.ProjectContextId != BuildEventContext.InvalidProjectContextId)
             {
-                errorEvent.ProjectFile = GetAndVerifyProjectFileFromContext(errorEvent.BuildEventContext);
+                errorEvent.ProjectFile = GetAndVerifyProjectFileFromContext(errorEvent, false);
             }
             else if ((messageEvent = buildEvent as BuildMessageEventArgs) != null && messageEvent.BuildEventContext != null && messageEvent.BuildEventContext.ProjectContextId != BuildEventContext.InvalidProjectContextId)
             {
-                messageEvent.ProjectFile = GetAndVerifyProjectFileFromContext(messageEvent.BuildEventContext);
+                // The AssemblyLoadBuildEventArgs are logged asynchronously, and build doesn't wait for those,
+                //  so it can happen that ProjectFinishedEventArgs occured first - removing the id->file mapping from map,
+                //  but AssemblyLoadsTracker still uses the BuildEventContext for that project
+                messageEvent.ProjectFile = GetAndVerifyProjectFileFromContext(messageEvent, buildEvent is AssemblyLoadBuildEventArgs);
             }
 
             if (OnlyLogCriticalEvents)
@@ -1927,14 +1930,17 @@ namespace Microsoft.Build.BackEnd.Logging
         /// <summary>
         /// Get the project name from a context ID. Throw an exception if it's not found.
         /// </summary>
-        private string GetAndVerifyProjectFileFromContext(BuildEventContext context)
+        private string GetAndVerifyProjectFileFromContext(BuildEventArgs eventArgs, bool allowCacheMiss)
         {
+            BuildEventContext context = eventArgs.BuildEventContext!;
             _projectFileMap.TryGetValue(context.ProjectContextId, out string projectFile);
 
             // PERF: Not using VerifyThrow to avoid boxing an int in the non-error case.
-            if (projectFile == null)
+            if (projectFile == null && !allowCacheMiss)
             {
-                ErrorUtilities.ThrowInternalError("ContextID {0} should have been in the ID-to-project file mapping but wasn't!", context.ProjectContextId);
+                ErrorUtilities.ThrowInternalError(
+                    "ContextID {0} should have been in the ID-to-project file mapping but wasn't! Encountered during logging message: '{1}'",
+                    context.ProjectContextId, eventArgs.Message);
             }
 
             return projectFile;

